File: Java/0003.java
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int left = 0; // Start of the window
        int maxLength = 0; // Result: the length of the longest substring without repeating characters

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            
            // If the character is already in the set, remove characters from the set 
            // until the duplicate character is removed
            while (set.contains(currentChar)) {
                set.remove(s.charAt(left));
                left++;
            }

            // Add the current character to the set
            set.add(currentChar);
            
            // Update the maximum length of substring found
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

File: Scala/0003.scala
object Solution {
    def lengthOfLongestSubstring(s: String): Int = {
        var maxLength = 0
        var start = 0
        val seen = scala.collection.mutable.Set[Char]()

        for (end <- s.indices) {
            while (seen.contains(s(end))) {
                seen.remove(s(start))
                start += 1
            }
            seen.add(s(end))
            maxLength = math.max(maxLength, end - start + 1)
        }

        maxLength
    }
}
